content/en/blog/_posts/2022-11-21-devicemanager-ga.md/index.md

```diff
@@ -0,0 +1,95 @@
+---
+layout: blog
+title: 'Kubernetes 1.26: Graduation of Device Manager to GA!'
+date: 2021-11-21
+slug: graduation-of-devicemanager-to-GA
+---
+
+**Author:** Swati Sehgal (Red Hat)
+
+##  Quick Intro to the Device Plugin framework
+
+### Introduction
+Device Plugin framework was introduced in Kubernetes v1.8 release as a vendor
+independent framework to enable discovery, advertisement and allocation of external
+devices without modifying core Kubernetes. The feature graduated to Beta in v1.10.
+Due to its widespread use and adoption, this feature is being graduated to GA in
+v1.26.
+
+Device Manager in Kubelet facilitates communication with device plugins over gRPC
+through Unix sockets. Device Manager and Device plugins both act as gRPC servers
+and clients by serving and connecting to the exposed gRPC services respectively.
+Device plugins serve a gRPC service that Kubelet connects to for device discovery,
+advertisement (as extended resources) and allocation. Device Manager connects to
+the `Registration` gRPC service served by Kubelet to register itself with Kubelet.
+
+### Pod requesting a device
+Please refer to the documentation for an [example](/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#example-pod) on how a pod can request a device exposed to the cluster by a device plugin.
+
+### Example Device Plugin implmentations 
+Here are some device plugins for example implmentations:
+- [NVIDIA device plugin for Kubernetes](https://github.com/NVIDIA/k8s-device-plugin)
+- [Collection of Intel device plugins for Kubernetes](https://github.com/intel/intel-device-plugins-for-kubernetes)
+- [SRIOV network device plugin for Kubernetes](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin)
+- [AMD GPU device plugin](https://github.com/RadeonOpenCompute/k8s-device-plugin)
+
+## Noteworthy developments since Device Plugin framework introduction
+
+### Kubelet APIs moved to kubelet staging repo
+External facing `deviceplugin` API packages moved from `k8s.io/kubernetes/pkg/kubelet/apis/`
+to `k8s.io/kubelet/pkg/apis/` in v1.17. Refer to [Move external facing kubelet apis to staging](https://github.com/kubernetes/kubernetes/pull/83551) for more details on the rationale behind this change.
+
+### Device Plugin API updates
+Additional gRPC endpoints introduced:
+  1. `GetDevicePluginOptions` is used by device plugins to communicate
+     options to the `DeviceManager` in order to indicate if `PreStartContainer`
+     or `GetPreferredAllocation` calls are supported and can be called before
+     making devices available to the container.
+  1. `GetPreferredAllocation` allows a device plugin to forward allocation
+     preferrence to the `DeviceManager` so it can incorporate this information
```

Comment:
I'd explictely call out that (IIRC) we are talking about NUMA-locality for preferred allocation

Reply:
My understanding of `GetPrefferedAllocation` endpoint is that once the set of devices suitable for assignment based on NUMA-locality have been identified, this endpoint gives a chance to the device plugin to weigh in on the final allocation decision. The preference can be NUMA-locality specific but doesn't have to be, it could be some device specific preference not related to NUMA. @klueska It would be great to get your input on this and in general on this blog post.

Reply:
I can be (partially) wrong indeed. What I'm trying to suggest is to be a bit more concrete re: what the preference entails, again IMO no need for lengthy description, just few words to set the direction.

Reply:
Ack, will update. Thanks